Crash Restoring State

From #90:
> I used the code you provided in my MyExpandableAdapter.java

>Everything works fine but after i browse the application, change the device configuration to landscape and switch to another application there was an error occurred related to 
ArrayIndexOutOfBound in the mExpandableAdapter.onRestoreInstanceState(savedInstanceState);

>Caused by: java.lang.IndexOutOfBoundsException: Invalid index 20, size is 20 
at com.bignerdranch.expandablerecyclerview.Adapter.ExpandableRecyclerAdapter.getListItem(ExpandableRecyclerAdapter.java:490)
at com.bignerdranch.expandablerecyclerview.Adapter.ExpandableRecyclerAdapter.onRestoreInstanceState(ExpandableRecyclerAdapter.java:445)
at com.eyey.android.eyeynews.ui.MainActivity.onRestoreInstanceState(MainActivity.java:87)

Also from #90 
Question:
>does the crash still occur if you don't have the code I provided? it seems like this issue may be separate from the changes I suggest.

Reply:
>I just tested it again, thing a bit different, When I put your code things work fine even i change device configuration and browse around however after a while of running by changing back and forth the applications. The same error occurred.

>When i don't use your code, things work fine.

>I think the out of bound part was the problem. 
Caused by: java.lang.IndexOutOfBoundsException: Invalid index 20, size is 20
There is only 20 objects, and I suppose the start index is 0.
and why in onRestoreInstanceState method the getListItem method try to access the object 20th.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,7 @@
+Version 2.0.4 (11/20/15)
+----------------------------
+- Fix crash when restoring state when an initially expanded row has been collapsed
+
 Version 2.0.3 (10/14/15)
 ----------------------------
 - Change `onListItemExpanded` and `onListItemCollapsed` to be called after list is updated, fixes bug in issue #112

--- a/README.md
+++ b/README.md
@@ -9,9 +9,9 @@
 ##Project Setup
 **Gradle**
 
-The [latest release](https://github.com/bignerdranch/expandable-recycler-view/releases/tag/v2.0.3) can be used by adding the following to your app's build.gradle:
+The [latest release](https://github.com/bignerdranch/expandable-recycler-view/releases/tag/v2.0.4) can be used by adding the following to your app's build.gradle:
 ```gradle
-compile 'com.bignerdranch.android:expandablerecyclerview:2.0.3'
+compile 'com.bignerdranch.android:expandablerecyclerview:2.0.4'
 ```
 
 ## Overview

--- a/app/src/main/java/com/ryanbrooks/expandablerecyclerviewsample/linear/horizontal/HorizontalChild.java
+++ b/app/src/main/java/com/ryanbrooks/expandablerecyclerviewsample/linear/horizontal/HorizontalChild.java
@@ -1,12 +1,14 @@
 package com.ryanbrooks.expandablerecyclerviewsample.linear.horizontal;
 
+import java.io.Serializable;
+
 /**
  * Custom child list item.
  *
  * This is for demo purposes, although it is recommended having a separate
  * child from your parent.
  */
-public class HorizontalChild {
+public class HorizontalChild implements Serializable {
 
     private String mChildText;
 

--- a/app/src/main/java/com/ryanbrooks/expandablerecyclerviewsample/linear/horizontal/HorizontalChild.java
+++ b/app/src/main/java/com/ryanbrooks/expandablerecyclerviewsample/linear/horizontal/HorizontalChild.java
@@ -7,6 +7,10 @@ import java.io.Serializable;
  *
  * This is for demo purposes, although it is recommended having a separate
  * child from your parent.
+ *
+ * You do not need to implement Serializable in order to use ExpandableRecyclerView. We are doing so
+ * here in order to store away our list of items (which you usually will have in a singleton or
+ * database)
  */
 public class HorizontalChild implements Serializable {
 

--- a/app/src/main/java/com/ryanbrooks/expandablerecyclerviewsample/linear/horizontal/HorizontalParent.java
+++ b/app/src/main/java/com/ryanbrooks/expandablerecyclerviewsample/linear/horizontal/HorizontalParent.java
@@ -10,6 +10,10 @@ import java.util.List;
  * Custom parent list item that holds a string and an int for displaying data in the parent item. You
  * can use any Object here as long as it implements ParentListItem and sets the list to a private
  * variable.
+ *
+ * You do not need to implement Serializable in order to use ExpandableRecyclerView. We are doing so
+ * here in order to store away our list of items (which you usually will have in a singleton or
+ * database)
  */
 public class HorizontalParent implements ParentListItem, Serializable {
 

--- a/app/src/main/java/com/ryanbrooks/expandablerecyclerviewsample/linear/horizontal/HorizontalParent.java
+++ b/app/src/main/java/com/ryanbrooks/expandablerecyclerviewsample/linear/horizontal/HorizontalParent.java
@@ -3,6 +3,7 @@ package com.ryanbrooks.expandablerecyclerviewsample.linear.horizontal;
 
 import com.bignerdranch.expandablerecyclerview.Model.ParentListItem;
 
+import java.io.Serializable;
 import java.util.List;
 
 /**
@@ -10,7 +11,7 @@ import java.util.List;
  * can use any Object here as long as it implements ParentListItem and sets the list to a private
  * variable.
  */
-public class HorizontalParent implements ParentListItem {
+public class HorizontalParent implements ParentListItem, Serializable {
 
     private List<HorizontalChild> mChildItemList;
     private String mParentText;

--- a/expandablerecyclerview/src/main/java/com/bignerdranch/expandablerecyclerview/Adapter/ExpandableRecyclerAdapter.java
+++ b/expandablerecyclerview/src/main/java/com/bignerdranch/expandablerecyclerview/Adapter/ExpandableRecyclerAdapter.java
@@ -435,48 +435,34 @@ public abstract class ExpandableRecyclerAdapter<PVH extends ParentViewHolder, CV
         }
 
         HashMap<Integer, Boolean> expandedStateMap = (HashMap<Integer, Boolean>) savedInstanceState.getSerializable(EXPANDED_STATE_MAP);
-        int fullCount = 0;
-        int childCount = 0;
-        Object listItem;
+        if (expandedStateMap == null) {
+            return;
+        }
+
+        List<Object> parentWrapperList = new ArrayList<>();
+        ParentListItem parentListItem;
         ParentWrapper parentWrapper;
-        List<?> childItemList;
-        int listItemCount = mItemList.size();
-        while (fullCount < listItemCount) {
-            listItem = getListItem(fullCount);
 
-            if (listItem instanceof ParentWrapper) {
-                parentWrapper = (ParentWrapper) listItem;
+        int parentListItemCount = mParentItemList.size();
+        for (int i = 0; i < parentListItemCount; i++) {
+            parentListItem = mParentItemList.get(i);
+            parentWrapper = new ParentWrapper(parentListItem);
+            parentWrapperList.add(parentWrapper);
 
-                if (expandedStateMap.containsKey(fullCount - childCount)) {
-                    parentWrapper.setExpanded(expandedStateMap.get(fullCount - childCount));
+            if (expandedStateMap.containsKey(i)) {
+                boolean expanded = expandedStateMap.get(i);
+                if (expanded) {
+                    parentWrapper.setExpanded(true);
 
-
-                    if (parentWrapper.isExpanded() && !parentWrapper.isInitiallyExpanded()) {
-                        childItemList = parentWrapper.getChildItemList();
-
-                        if (childItemList != null) {
-                            int childListItemCount = childItemList.size();
-                            for (int j = 0; j < childListItemCount; j++) {
-                                fullCount++;
-                                childCount++;
-                                mItemList.add(fullCount, childItemList.get(j));
-                            }
-                        }
-                    } else if (!parentWrapper.isExpanded() && parentWrapper.isInitiallyExpanded()) {
-                        childItemList = parentWrapper.getChildItemList();
-                        int childListItemCount = childItemList.size();
-                        for (int j = 0; j < childListItemCount; j++) {
-                            mItemList.remove(fullCount + 1);
-                            listItemCount--;
-                        }
+                    int childListItemCount = parentWrapper.getChildItemList().size();
+                    for (int j = 0; j < childListItemCount; j++) {
+                        parentWrapperList.add(parentWrapper.getChildItemList().get(j));
                     }
                 }
-            } else {
-                childCount++;
             }
-
-            fullCount++;
         }
+
+        mItemList = parentWrapperList;
 
         notifyDataSetChanged();
     }

--- a/expandablerecyclerview/src/main/java/com/bignerdranch/expandablerecyclerview/Adapter/ExpandableRecyclerAdapter.java
+++ b/expandablerecyclerview/src/main/java/com/bignerdranch/expandablerecyclerview/Adapter/ExpandableRecyclerAdapter.java
@@ -467,6 +467,7 @@ public abstract class ExpandableRecyclerAdapter<PVH extends ParentViewHolder, CV
                         int childListItemCount = childItemList.size();
                         for (int j = 0; j < childListItemCount; j++) {
                             mItemList.remove(fullCount + 1);
+                            listItemCount--;
                         }
                     }
                 }

--- a/gradle.properties
+++ b/gradle.properties
@@ -1,5 +1,5 @@
-VERSION_NAME=2.0.3
-VERSION_CODE=5
+VERSION_NAME=2.0.4
+VERSION_CODE=6
 GROUP=com.bignerdranch.android
 
 POM_DESCRIPTION=A lightweight Android library that allows for the expansion and collapsing of RecyclerView items

